ls1intum/Artemis
# Feature/grading instructions/sgi feedback reference

### Checklist

### Motivation and Context


We can derive lots of useful data and statistics about the performance of the students from the assessment of their submissions. Beside the final total score we can also derive more data from each submission element and their assessment which resembles a partial assessment. A link between the partial assessment and the applied grading instruction would therefore be very beneficial. 
Through this link instructors can conclude which part of the assignment the students were struggling with or which part the student performed especially good at.
In addition, this link allows us to handle subsequent mistakes (Folge Fehler)
### Description

We save for each feedback element the id of the grading instruction applied on this feedback of a submission element.
We check the maxCountInScore value set by the instructors which resembles how often points for a single instruction should be added to the total score if applied multiple times.
If the number of application is higher than the number of maxCountInScore then the points are not added to the final score.
The default value 0 of maxCountInScore resemble that there is no limit of application, meaning that the instruction can be applied as often as possible.
### Changes to the Database:
In order to tie a grading instruction to a feedback element I added the column `grading_instruction_id` which saves the id of the grading instruction applied to the feedback element. One Grading instruction could be applied to many feedback elements therefore I added a one to many relation between grading_instruction and feedback table. (I followed here the same principle as the one implemented between Participation and Submission: one Participation <-> many submissions)
Another approach would be to save a physical reference string, like the one that refers the feedback to the text segment (in text exercises) or SVG (in modeling exercises). I didn't choose this approach because I thought it might be restrictive in the future if I want to automatically update all submitted assessments when the grading instructions are updated. 
![image](https://user-images.githubusercontent.com/50635005/84578488-170bca80-adc6-11ea-8142-0334d568ca0e.png)

--TODO--
adjust apollon to support this feature for modeling exercise assessment as well
### Steps for Testing


1.login as an instructor and create an file upload/text exercise with structured grading criteria. Make sure to sett at least for one instruction the maxCountInScore attribute with a value higher than 0
2. login as a student and submit solutions for these exercises you just created
3. login as a tutor and asses the students submissions
4. open the result after submit or override and check that the score is calculated and displayed correctly and subsequent mistakes are taken into consideration
5. check in general (for modeling and programming ex) that the score calculation still works correctly for all other exercise types

### Screenshots


![image](https://user-images.githubusercontent.com/50635005/83982090-e5a58180-a923-11ea-8d0a-890d1c3051e0.png)
![image](https://user-images.githubusercontent.com/50635005/83982110-108fd580-a924-11ea-9b12-127246c21ee2.png)
